releases-implementation/README.md

```diff
@@ -0,0 +1,415 @@
+
+Some companies may want to upgrade apps faster and more often than Kubernetes and operating system, because the former is less impactful to their workload than the latter and they do not require the latest Kubernetes and OS versions, but are satisfied with any supported and stable version. Others may not care about it and they would like to get newer versions of everything equally fast, or they want the latest Kubernetes releases to be deployed as soon as possible.
+
+To accommodate the needs of different types of businesses, we may need not one, but at least few different **release models**, so that customers can choose which one is working the best for their use case.
+
+In addition to all the above, even for a single release model, it may not be a good idea to make new releases available to all clusters at the same time, so it can be useful to have different **release channels**, where some release channels are getting new releases faster than the others. E.g. see [GKE release channels](https://cloud.google.com/kubernetes-engine/docs/concepts/release-channels). A development cluster may use a faster release channel, while a production cluster may use a slower one. Some companies may use faster release channel for all their clusters, while others may use a slower one similarly for all their workload.
+
+## 4. Proposal
+
+This section defines how releases are implemented, where they are, how they are developed, tested and delivered.
+
+Then it compares the proposed solution to the current one that we have for our new KaaS product that is based on Cluster API.
+
+### 4.1. Implementation of releases
+
+This document proposes to continue using `giantswarm/releases` repository for creating and delivering releases, albeit in a simplified way when compared to the vintage releases. Release resources are used in a minimal way, and cluster-$provider apps continue to be used for deploying workload clusters.
+
+Briefly put, cluster-$provider apps are still deployed in almost exactly same way, with the following few differences:
+- In cluster-$provider app manifest, instead of specifying `.spec.version`, we MUST specify `release.giantswarm.io/version` label.
+- Information about app version, catalog and dependencies MUST BE obtained from the Release resource (during Helm rendering phase) when deploying a production cluster.
```

Comment:
I assume this refers to needing to make the change in app-operator (or similar) to lookup the correct values.

Have you considered how this might be handled in the future when we migrate to using Flux and HelmReleases rather than app-operator?

Reply:
We should 100% avoid building anything custom into app operator for this, but AFAIK as long as these values are supplied through a configmap we should be fine no matter which helm rendering tool is used as they all support merging configs.

Reply:
I thought app operator will be there to manage `Apps` (transforming them into right HelmRelease objects offering the 3level config) meanwhile chart operator is going to dissapear

Reply:
> I assume this refers to needing to make the change in app-operator (or similar) to lookup the correct values.

No changes in app-operator are needed. Release CR is read in the Helm template with Helm `lookup` function, and then app and component versions are read and set in the templates.

> Have you considered how this might be handled in the future when we migrate to using Flux and HelmReleases rather than app-operator?

AFAIK Team Honeybadger is working on replacing chart-operator with Flux Helm controller, but app-operator is not going anywhere any time soon. Even if it did, that does not affect the work here in any way, as the change is just about Helm reading a Release CR, getting some value from it and putting that value to some field of some CR (e.g. App version in `.spec.version`, HelmRelease version in `.spec.chart.spec.version`, etc).